cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,342 @@
+
+2. The constraint expressions defined in the `REQUIRE` clauses of the constraint definition must all evaluate to `true`.
+
+3. [[domain-exception]]Entities for which a constraint expression evaluate to `null` under Cypher's ternary logic are _excluded_ from the constraint domain, even if they fit within the constraint pattern.
+
+==== Errors
+
+The following list describes the situations in which an error will be raised:
+
+* Attempting to create a constraint on a graph where the data does not comply with the constraint criterion.
+* Attempting to create a constraint with a name that already exists.
+* Attempting to drop a constraint referencing a non-existent name.
+* Attempting to modify the graph in such a way that it would violate a constraint.
+
+==== Mutability
+
+Once a constraint has been created, it may not be amended.
+Should a user wish to change its definition, it has to be dropped and recreated with an updated structure.
+
+[[uniqueness]]
+==== Uniqueness
+
+The new operator `UNIQUE` is only valid as part of a constraint predicate.
+It takes as argument a single property expression, and asserts that this property is unique across the domain of the constraint.
+Following on rule <<domain-exception,3.>> above, entities for which the property is not defined (is `null`) are not part of the constraint domain.
+
+.Example of a constraint definition using `UNIQUE`, over the domain of nodes labeled with `:Person`:
+[source, cypher]
+----
+CREATE CONSTRAINT only_one_person_per_name
+FOR (p:Person)
+REQUIRE UNIQUE p.name
+----
+
+[[primary-key]]
+==== Primary key
+
+The new operator `PRIMARY KEY` is only valid as part of a constraint predicate.
+It takes as argument one or more property expressions, and asserts that the combination of the evaluated values of the expressions (forming a tuple) is unique across the constraint domain.
+It further asserts that the property expressions all exist on the entities of the domain, and thus avoids applicability of rule <<domain-exception, 3.>> above. The domain of a primary key constraint is thus exactly defined as all entities which fit the constraint pattern.
+
+.Example of a constraint definition using `PRIMARY KEY`, over the domain of nodes labeled with `:Person`:
+[source, cypher]
+----
+CREATE CONSTRAINT person_details
+FOR (p:Person)
+REQUIRE PRIMARY KEY p.name, p.email, p.address
+----
+
+A semantically equivalent constraint is achieved by composing the use of the `UNIQUE` operator with `exists()` predicates, as exemplified by:
+
+.Example of a constraint definition equivalent to the above `PRIMARY KEY` example:
+[source, cypher]
+----
+CREATE CONSTRAINT person_details
+FOR (p:Person)
+REQUIRE UNIQUE p.name
+REQUIRE UNIQUE p.email
+REQUIRE UNIQUE p.address
+REQUIRE exists(p.name) AND exists(p.email) AND exists(p.address)
```

Comment:
This decomposition seems wrong to me, we would want the compound key to require that the _combination_ of `p.name`, `p.email`, and `p.address` is unique, not that each of them is unique on their own. But we do require that they each exist.